### PR TITLE
Add DataModel::Provider Invoke usage

### PR DESCRIPTION
**Pull request by andy31415**
### Changes

- Command handler processing has its own `AddStatus` (and variants) and may be processed asynchronously, so as a result we now have `Provider::Invoke` return a optional where missing means `status handled separately`
- Update IM to call the data model provider invoke implementation if data model provider processing is enabled
- Update unit tests

### CODE REVIEW

File: src/app/InteractionModelEngine.cpp [hunk 1]
```diff
@@ -1717,6 +1733,7 @@ void InteractionModelEngine::DispatchCommand(CommandHandlerImpl & apCommandObj,
     }
 
     DispatchSingleClusterCommand(aCommandPath, apPayload, &apCommandObj);
+#endif
```
Comment (tehampson) [inline src/app/InteractionModelEngine.cpp:1736]: nit,
```suggestion
#endif // CHIP_CONFIG_USE_DATA_MODEL_INTERFACE
```

File: src/app/data-model-provider/Provider.h [hunk 2]
```diff
@@ -98,8 +100,9 @@ class Provider : public ProviderMetadataTree
     ///    - data (as CommandDataIB) which is assumed a "response as a success"
     ///    - status (as a CommandStatusIB) which is considered a final status, usually an error however
     ///      cluster-specific success statuses also exist.
-    virtual ActionReturnStatus Invoke(const InvokeRequest & request, chip::TLV::TLVReader & input_arguments,
-                                      CommandHandler * handler) = 0;
+    ///
+    virtual std::optional<ActionReturnStatus> Invoke(const InvokeRequest & request, chip::TLV::TLVReader & input_arguments,
```
Comment (tehampson) [inline src/app/data-model-provider/Provider.h:104]: nit, can you update documentation? Please mention that if there is an error return that the handler has not sent anything in response to the command

File: src/app/InteractionModelEngine.cpp [hunk 3]
```diff
@@ -1699,6 +1703,18 @@ CHIP_ERROR InteractionModelEngine::PushFront(SingleLinkedListNode<T> *& aObjectL
...
                                              TLV::TLVReader & apPayload)
 {
+#if CHIP_CONFIG_USE_DATA_MODEL_INTERFACE
+
+    DataModel::InvokeRequest request;
+    request.path = aCommandPath;
+
+    std::optional<DataModel::ActionReturnStatus> status = GetDataModelProvider()->Invoke(request, apPayload, &apCommandObj);
+
+    if (status.has_value())
```
Comment (tehampson) [inline src/app/InteractionModelEngine.cpp:1713]: What is value() is CHIP_NO_ERROR (the equivalent that means OK for ActionReturnStatus) ?

**Review by lpbeliveau-silabs — APPROVED** [review]

File: src/app/InteractionModelEngine.cpp (hunk 3 above)
Comment (andy31415) [reply src/app/InteractionModelEngine.cpp:1713]: Updated documentation some more. std::nullopt is distinctly different from CHIP_NO_ERROR. Added notes both here and in the Provider.h documentation.

File: src/app/data-model-provider/Provider.h [hunk 4]
```diff
@@ -88,18 +88,25 @@ class Provider : public ProviderMetadataTree
...
     /// `handler` is used to send back the reply.
+    ///    - returning `std::nullopt` means that return value was placed in handler directly.
+    ///      This includes cases where command handling and value return will be done asynchronously.
     ///    - returning a value other than Success implies an error reply (error and data are mutually exclusive)
     ///
-    /// Returning anything other than CHIP_NO_ERROR or Status::Success (i.e. success without a return code)
-    /// means that the invoke will be considered to be returning the given path-specific status WITHOUT any data (any data
-    /// that was sent via CommandHandler is to be rolled back/discarded).
+    /// Return value expectations:
+    ///   - if a response has been placed into a `handler` then std::nullopt MUST be returned. In particular
```
Comment (bzbarsky-apple) [inline src/app/data-model-provider/Provider.h:96]:
```suggestion
    ///   - if a response has been placed into `handler` then std::nullopt MUST be returned. In particular
```

File: src/app/data-model-provider/Provider.h [hunk 5]
```diff
@@ -88,18 +88,25 @@ class Provider : public ProviderMetadataTree
...
+    ///     note that CHIP_NO_ERROR is NOT the same as std::nullopt:
+    ///        > CHIP_NO_ERROR means handler had no status set and we caller to AddStatus(success)
```
Comment (bzbarsky-apple) [inline src/app/data-model-provider/Provider.h:98]:
```suggestion
    ///        > CHIP_NO_ERROR means handler had no status set and we expect the caller to AddStatus(success)
```

File: src/app/data-model-provider/Provider.h [hunk 6]
```diff
@@ -88,18 +88,25 @@ class Provider : public ProviderMetadataTree
...
+    ///     convenience to make writing Invoke calls easier.
     ///
-    /// This is because only one of the following may be encoded in a response:
-    ///    - data (as CommandDataIB) which is assumed a "response as a success"
-    ///    - status (as a CommandStatusIB) which is considered a final status, usually an error however
-    ///      cluster-specific success statuses also exist.
-    virtual ActionReturnStatus Invoke(const InvokeRequest & request, chip::TLV::TLVReader & input_arguments,
-                                      CommandHandler * handler) = 0;
+    /// Expectations if invoke returns data:
+    ///   - if a command data is to be returned, the status can only be 'Success' (it cannot be cluster
```
Comment (bzbarsky-apple) [inline src/app/data-model-provider/Provider.h:105]: This documentation does not make sense.  If command data is to be returned, doesn't the status have to be nullopt?

Comment (andy31415) [reply src/app/data-model-provider/Provider.h:105]: I deleted that part - it used to be there as a note that `cluster specific success` is incompatible with `data` (i.e. data has bare success implied). However this is rexplained now above about return data expectations.

**Review by tehampson — APPROVED** [review]

**Pull request merged**